### PR TITLE
Replace Variable.volatile with torch.no_grad()

### LINKED ISSUE #3627: Proposal: replace Variable.volatile with global switch
Variable.volatile forces outputs to not require gradients if any of the inputs are marked volatile. This works OK in the forward pass, but we're forced to [change the meaning](https://github.com/pytorch/pytorch/blob/b06c59e543aa26586087c19fb7b713f8872105bb/torch/csrc/autograd/functions/accumulate_grad.cpp#L38) in the backwards. Gradients are sometimes volatile and sometimes not, which is awkward if you add them back to parameters, such as in optimizers.

We should replace `volatile` with a context manager in Python. (Chainer already did this with [`no_backprop_mode()`](http://docs.chainer.org/en/stable/reference/core/generated/chainer.no_backprop_mode.html#chainer.no_backprop_mode)).

At the C++ level, we should replace it with a thread-local global switch.

This will simplify the logic in the backwards: by default `backwards()` will set "no-backprop mode", unless `create_graph` is True.

### PR DESCRIPTION
**Pull request by colesbury**
This removes volatile from Variable. The functionality is mostly
replaced by a global (thread-local) flag, which is controlled by
torch.set_grad_enabled() and the context manager torch.no_grad().

In C++, the flag is exposed through `GradMode::is_enabled()` and `GradMode::set_enabled()`

Fixes #3627

### CODE REVIEW

**Comment by colesbury** [comment]
I placed the context manage in the `torch` package instead of `torch.autograd` in anticipation of merging `Tensor` and `Variable`.

**Comment by apaszke** [comment]
Haven't started the review, but can we please call it `no_grad` or sth like this? Backprop is very nn-specific and is a special case of a more general reverse-mode AD, which is the level on which autograd operates.

**Comment by Randl** [comment]
ping

File: torch/csrc/autograd/variable.cpp [hunk 1]
```diff
@@ -107,19 +105,8 @@ VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_, VarFlags fl
...
-        "requires_grad is False and base.requires_grad is True. Cannot use "
-        "this view in a differentiable operation. Re-create the view from its "
-        "base Variable after the last in-place modification.");
-  }
-  if (base.is_volatile() && !is_volatile) {
-    throw std::runtime_error(
-        "is_volatile is False and base.is_volatile is True. Cannot use "
-        "this view in a differentiable operation. Re-create the view from its "
-        "base Variable after the last in-place modification.");
+  if (!_grad_fn && !base.requires_grad()) {
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: I think this breaks when you `detach_()` the base. I think we're end up with a view that doesn't require grad, but the version counter is the same, so views won't even notice and will happily return their old grad_fns. We need to check `self.grad_fn()`

**Review by apaszke — COMMENTED** [review]
Looks great. I have a ton of questions, because I want to make sure that everything works fine, and we haven't missed anything. Also, I think there are a few things that should be fixed before we merge this.

File: torch/csrc/autograd/variable.h [hunk 2]
```diff
@@ -169,30 +153,31 @@ struct VariableViewImpl : public VariableImpl {
...
-Variable make_variable(at::Tensor data, std::shared_ptr<Function> grad_fn);
-
-inline Variable make_variable(at::Tensor data, bool requires_grad, bool is_volatile=false) {
-  return make_variable(std::move(data), VarFlags(requires_grad, is_volatile));
+inline Variable make_variable(at::Tensor data, int output_nr, std::shared_ptr<Function> grad_fn) {
+  if (data.defined() && data.dim() == 0) {
+    // don't expose 0-dim tensors to Variable API.
+    data = data.as_strided_({1}, {1});
+  }
+  return Variable(new VariableImpl(std::move(data), false, output_nr, std::move(grad_fn)), false);
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: It makes the `requires_grad()` computation a little simpler (we don't have to explicitly check if the variable is a leaf).

File: torch/csrc/autograd/variable.h [hunk 3]
```diff
@@ -273,11 +258,8 @@ inline int& Variable::output_nr() {
   return get()->output_nr;
 }
 
-inline const bool& Variable::requires_grad() const {
-  return get()->requires_grad;
-}
-inline bool& Variable::requires_grad() {
-  return get()->requires_grad;
+inline bool Variable::requires_grad() const {
+  return get()->_requires_grad || get()->_grad_fn || (is_view() && base().requires_grad());
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: I've updated the comment.

File: torch/nn/modules/module.py [hunk 4]
```diff
@@ -652,11 +652,8 @@ def zero_grad(self):
         """Sets gradients of all model parameters to zero."""
         for p in self.parameters():
             if p.grad is not None:
-                if p.grad.volatile:
-                    p.grad.data.zero_()
-                else:
-                    data = p.grad.data
-                    p.grad = Variable(data.new().resize_as_(data).zero_())
+                p.grad.detach_()
+                p.grad.data.zero_()
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/nn/modules/module.py]: I will change this to `p.grad.zero_()` so that the version counter is incremented.

File: torch/optim/optimizer.py [hunk 5]
```diff
@@ -137,11 +137,8 @@ def zero_grad(self):
         for group in self.param_groups:
             for p in group['params']:
                 if p.grad is not None:
-                    if p.grad.volatile:
-                        p.grad.data.zero_()
-                    else:
-                        data = p.grad.data
-                        p.grad = Variable(data.new().resize_as_(data).zero_())
+                    p.grad.detach_()
+                    p.grad.zero_()
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/optim/optimizer.py]: This increments the version counter.

File: torch/csrc/jit/python_compiled_function.cpp [hunk 6]
```diff
@@ -130,8 +130,7 @@ struct CompiledFunction {
   TraceForKey& getTrace(ParsedArgs& args) {
     auto it = ktraces_.find(args.desc);
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/jit/python_compiled_function.cpp]: I'll add `BackpropMode::is_enabled()` as a key (replacing `is_volatile_`)

File: torch/csrc/autograd/variable.cpp [hunk 7]
```diff
@@ -137,26 +124,17 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
...
-  TORCH_ASSERT(flags.requires_grad == bool(grad_fn));
-  if (grad_fn) {
-    TORCH_ASSERTM(grad_fn->num_inputs == 1, "Functions which modify views in-place must return a single Variable");
-  } else {
-    // TODO: perhaps we should enable this case by setting base.requires_grad=False
-    // and base.grad_fn = nullptr.
-    TORCH_ASSERTM(!base.requires_grad(), "base.requires_grad does not match view.requires_grad");
+void VariableViewImpl::rebase_history(int output_nr, std::shared_ptr<Function> grad_fn) {
+  if (!grad_fn) {
+    return;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: Our grad_fn may be defined, in which case we don't modify it.

File: torch/csrc/autograd/variable.h [hunk 8]
```diff
@@ -58,8 +48,8 @@ struct Variable : public at::Tensor {
 
   inline const std::shared_ptr<Function>& grad_fn() const;
 
-  // Updates the flags and grad_fn of an existing Variable. Called after in-place modifications.
-  inline void rebase_history(VarFlags flags, int output_nr, std::shared_ptr<Function> grad_fn);
+  // Updates the grad_fn of an existing Variable. Called after in-place modifications.
+  inline void rebase_history(int output_nr, std::shared_ptr<Function> grad_fn);
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: Yeah, it's unnecessary because `requires_grad == (grad_fn != nullptr)`. I'm going to update it so that it's only called when grad_fn is not null.

File: torch/csrc/autograd/variable.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: Actually, I'll change this so that rebase_history is only called if `grad_fn` is not null. I think that might be a little more clear.

File: torch/csrc/autograd/variable.cpp [hunk 9]
```diff
@@ -137,26 +124,17 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
...
+  TORCH_ASSERTM(grad_fn->num_inputs == 1, "Functions which modify views in-place must return a single Variable");
   this->output_nr = output_nr;
-  base.requires_grad() |= flags.requires_grad;
-  base.is_volatile() |= flags.is_volatile;
-  if (grad_fn) {
-    base.output_nr() = 0;
-    base.get()->_grad_fn = std::make_shared<CopySlices>(
-        base, TensorGeometry(data), std::move(grad_fn));
-  }
+  this->_grad_fn = grad_fn;
```
[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: The value of `this->_grad_fn` on the view doesn't matter here. It'll be overwritten by the next call to `grad_fn()` to an instance of `AsStridedBackward` and look like your third graph.

I'll change this so it just nulls out `_grad_fn`

File: torch/csrc/autograd/variable.cpp (hunk 1 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: Not sure what you expect here: if you modify Tensors through a detached variable it's not going to affect gradients. That was true before and after this PR.

File: test/test_autograd.py [hunk 10]
```diff
@@ -1551,7 +1535,7 @@ def backward(ctx, grad_output):
 
         x = Variable(torch.randn(2, 2), requires_grad=True)
         out = Reenter.apply(x)
-        out.sum().backward()
+        out.sum().backward(create_graph=True)
```
[the remark this answers is not in the document]
Comment (apaszke) [reply test/test_autograd.py]: The line numbers have changed, but I now see what's the problem. Can we add `torch.with_grad` so that such ops can ensure that they always run in a differentiable context, even when we don't instantiate the graph for backward (which shouldn't be necessary)

File: test/test_autograd.py [hunk 11]
```diff
@@ -146,31 +144,30 @@ def backward(ctx, grad):
...
+        x_grad, x_grad_clone = compute_grad(create_graph=True)
+        self.assertEqual(x_grad, x_grad_clone)
 
     def test_hessian_vector(self):
         x = Variable(torch.randn(2, 2), requires_grad=True)
         y = Variable(torch.randn(2, 2), requires_grad=True)
 
         z = x ** 2 + y * x + y ** 2
-        z.backward(Variable(torch.ones(2, 2), requires_grad=True), retain_graph=True)
+        z.backward(torch.ones(2, 2), create_graph=True)
```
Comment (apaszke) [inline test/test_autograd.py]: Is it now the case that passing a Variable that requires grad is not sufficient to trigger creation of the graph? If yes, we should at least print a warning

File: tools/autograd/gen_variable_type.py [hunk 12]
```diff
@@ -870,16 +870,19 @@ def mk_tuple_getters(pred):
...
         env['version_counter'] = ''
         env['no_zero_dim'] = ''
+        env['is_view'] = str(is_view).lower()
         base_call = BASE_CALL.substitute(combined)
         if declaration['inplace']:
             env['check_inplace'] = 'check_inplace(self);'
             env['version_counter'] = 'increment_version(self);'
-            modifies_data = 'false' if is_view else 'true'
-            env['set_flags'] = SET_FLAGS_INPLACE.substitute(combined, modifies_data=modifies_data)
+            if is_view:
```
[the remark this answers is not in the document]
Comment (apaszke) [reply tools/autograd/gen_variable_type.py]: Can you please add a comment saying this? It's not obvious

File: setup.py [hunk 13]
```diff
@@ -456,6 +456,7 @@ def check_file(f):
     "torch/csrc/jit/passes/onnx/peephole.cpp",
     "torch/csrc/jit/generated/aten_dispatch.cpp",
     "torch/csrc/autograd/init.cpp",
+    "torch/csrc/autograd/backprop_mode.cpp",
```
[the remark this answers is not in the document]
Comment (apaszke) [reply setup.py]: I agree. I just wanted to make sure it works this way

File: torch/autograd/backprop_mode.py [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+    """
+
+    def __init__(self):
+        self.prev = torch.is_backprop_enabled()
+
+    def __enter__(self):
+        torch.set_backprop_enabled(False)
+
+    def __exit__(self, *args):
+        torch.set_backprop_enabled(self.prev)
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/autograd/backprop_mode.py]: @soumith why?

File: torch/csrc/autograd/python_variable.cpp [hunk 15]
```diff
@@ -389,7 +358,7 @@ int THPVariable_set_requires_grad(THPVariable *self, PyObject *obj)
     THPUtils_setError("you can only change requires_grad flags of leaf variables.%s", hint);
     return -1;
   }
-  var.requires_grad() = (obj == Py_True);
+  var.get()->_requires_grad = (obj == Py_True);
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/python_variable.cpp]: That's fine. Let's leave it as is

File: torch/csrc/autograd/functions/accumulate_grad.cpp [hunk 16]
```diff
@@ -34,31 +35,22 @@ auto AccumulateGrad::apply(const variable_list& grads) -> variable_list {
...
-  // TODO: Currently if var.grad is volatile and new-grad is non-volatile we
-  // accumulate in-place. We should reconsider this and perhaps add the
-  // gradients out-of-place.
-
   auto& grad = variable.grad();
   if (!grad.defined()) {
-    grad = apply_fn<Clone>()(new_grad);
-  } else if (grad.is_volatile()) {
+    variable.grad() = new_grad.clone();
+  } else if (!BackpropMode::is_enabled()) {
```
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/functions/accumulate_grad.cpp]: It wouldn't be too bad if you did in-place autograd ops. My point is that you're doing in-place tensor ops, so you're opting out of all kinds of correctness checks we have and will start getting silent errors

File: torch/csrc/autograd/variable.cpp (hunk 1 above)
Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: I can't see how is this related to what I said. My point is that doing `detach_()` on the base will also cause the views to not require grad, but they will retain their grad functions

File: torch/csrc/autograd/variable.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: Yes, that would definitely help

File: torch/csrc/autograd/variable.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: Immediate update sounds like a better idea. The thing we're doing now is really weird

File: torch/csrc/autograd/variable.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (apaszke) [reply torch/csrc/autograd/variable.h]: Sorry, can you please elaborate on this. I'm not sure what you meant. Is it because `_requires_grad` **should be** set to `false` if the variable is not a leaf?

[the remark this answers is not in the document]
Comment (colesbury) [reply torch/csrc/autograd/variable.h]: I'll add an assertion, but the public API (`make_variable`) only takes `requires_grad` or only takes `grad_fn` (not both).

File: torch/csrc/autograd/variable.cpp (hunk 1 above)
Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: Detaching the base does not detach views.  Any variable with a non-null `grad_fn` also has `var.requires_grad() = true`:

https://github.com/colesbury/pytorch/blob/d5fa2c1e8af3471fbcd9ccb8d4a125036cd31707/torch/csrc/autograd/variable.h#L260-L262

**Review by apaszke — COMMENTED** [review]
Can you please add a test that checks that views of a detached base still require grad and have grad functions?

File: torch/csrc/autograd/variable.cpp [hunk 17]
```diff
@@ -137,26 +125,15 @@ std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
...
-  base.is_volatile() |= flags.is_volatile;
-  if (grad_fn) {
-    base.output_nr() = 0;
-    base.get()->_grad_fn = std::make_shared<CopySlices>(
-        base, TensorGeometry(data), std::move(grad_fn));
-  }
+  base.output_nr() = 0;
+  base.get()->_grad_fn = std::make_shared<CopySlices>(
+      base, TensorGeometry(data), std::move(grad_fn));
+  get_grad_fn();  // trigger an update to the view's grad_fn
```
Comment (apaszke) [inline torch/csrc/autograd/variable.cpp]: There's one non-documented invariant - setting the flags needs to happen **after** we increment the version counter, or we won't trigger an update here. Would be nice to mention this next to the template in the code gen.

Comment (apaszke) [reply torch/csrc/autograd/variable.cpp]: See #4272

Comment (colesbury) [reply torch/csrc/autograd/variable.cpp]: You won't get the immediate update to `grad_fn` but it will still be correct if you swap the order. The next time `grad_fn()` is accessed it will update the function.

**Comment by rawalkhirodkar** [comment]
Can we update the documentation highlighting the removal of volatile in favor of torch.no_grad( )?
Thank you

**Comment by apaszke** [comment]
@rawalkhirodkar yes, that's definitely needed. Can you please send a PR?

**Comment by Randl** [comment]
Also I'm not sure what to do if only part of variables don't require grads

**Comment by apaszke** [comment]
@Randl `requires_grad` still works as it used to (when grad mode is enabled)

**Comment by Randl** [comment]
@apaszke I probably didn't explain myself well. 

As far as I understand, `with torch.no_grad()` is equivalent to running with **all** variables volatile. What if want only one variable to be volatile?

**Comment by apaszke** [comment]
Then you need to separate the volatile codepaths from the other ones. I haven't yet seen a case where keeping only some Variables volatile is useful.

**Comment by fmassa** [comment]
@Randl also note that volatile variables propagate in the graph. So if you have 2 variables, `a` and `b`, where only `a` is volatile, the result of an operation between `a` and `b` will be volatile as well.

**Comment by my-hello-world** [comment]
hello,i want to know that if  PyTorch 0.4 can with CUDA7.5 support.cause when i install cuda8.0,it show:"    Found GPU1 Quadro K2000 which is of cuda capability 3.0.
PyTorch no longer supports this GPU because it is too old."
i sorry that i can't  afford the new Nvidia.
now, i install cuda7.5,and when i try "export CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" # [anaconda root directory]
conda install numpy pyyaml mkl mkl-include setuptools cmake cffi typing
conda install -c mingfeima mkldnn
conda install -c pytorch magma-cuda75
git clone --recursive https://github.com/pytorch/pytorch
cd pytorch
python setup.py install"   the pytorch -V is 0.3.0
 so how can i do?
My environment is : ubuntu14.04,anaconda, python3.6.5 
Thanks!!!!!
